modules/grandpa/src/benchmarking.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Benchmarks for the GRANDPA Pallet.
+
+// Braindump for benches:
+//
+// Will want to make sure number of requests is low
+// Will want to Make sure we have already ImportedHeaders and BestFinalized
+//		(E.g pallet is initialized)
+// Will want to check how heavy authority set changes are
+//		How does weight increase as the authority list size increases since we have to
+//		store that info in the pallet
+//	Most important thing will be checking how justification verification scales
+//		It's checking ancestry proofs, so we'll need to create valid, but different sized
+//		justifications
+
+use crate::*;
+
+use bp_test_utils::{
+	authority_list, keyring, make_justification_for_header, JustificationGeneratorParams, Keyring::*,
+	TEST_GRANDPA_ROUND, TEST_GRANDPA_SET_ID,
+};
+use frame_benchmarking::{benchmarks_instance_pallet, whitelisted_caller};
+use frame_system::RawOrigin;
+use sp_finality_grandpa::AuthorityId;
+use sp_runtime::traits::{One, Zero};
+use sp_std::vec;
+
+pub trait Config<I: 'static = ()>: crate::Config<I> {
+	// We need some way for the benchmarks to use headers in a "generic" way. However, since we do
+	// use a real runtime we need a way for the runtime to tell us what the concrete type is.
+	fn bridged_header(num: BridgedBlockNumber<Self, I>) -> BridgedHeader<Self, I>;
+}
+
+benchmarks_instance_pallet! {
+	// What we want to check here is the effect of vote ancestries on justification verification
+	// time. We will do this by varying the number of ancestors our finality target has.
+	submit_finality_proof_on_single_fork {
+		let n in 1..10;
```

Comment:
The difference between 1 and 10 headers is 9 hash operations and BTreeMap inserts, right? So probably you'll want more headers to see actual difference. Probably you'll even get conflicting results here (e.g. time of bench with depth=10 will be lesser than time of bench with depth=1 e.g. because of context switches or similar). So if the point of this benchmark is to build fn(depth) -> time, then it makes sense to make it much larger.